functions.md

```diff
@@ -0,0 +1,302 @@
+
+## Naming and overriding
+
+Function declarations must have unique names. E.g.,
+
+```
+fn foo() {...}      // ok
+fn bar() {...}      // ok
+fn bar(): num {...} // error: duplicate names
+```
+
+Exception: functions may have the same name if they have fully distinct `self` types. A function with no self argument is considered distinct to any functions with `self`. If the `self` types are unrelated, there is no restriction on the other arguments or return type. If the `self` types are subtypes, other arguments must match exactly[^contra] and return types must be covariant. Due to the `self`'array sugar, `T` is not considered distinct from `[T]` or `[T; 1]`.
+
+E.g., (note I use a 'top type' called `Any` in these examples, but I'm not actually proposing we have such a type, it's just easy to explain with).
+
+```
+fn foo(a: bool) { ... }
+fn foo(self: num) { ... }           // ok
+fn foo(self: T1): T1 { ... }        // ok
+fn foo(self: T2): T2 { ... }        // ok
+fn foo(self: [num]) { ... }         // error self: num and [num]
+fn foo(self: Any, a?: num) { ... }  // error args must match, return types must match
+fn foo(self: Any): Any { ... }      // ok
+```
+
+E.g., point constructors (similarly for `vec`/`Vec2d`/`Vec3d`):
+
+```
+fn pt(self: [num; 2]): Point2d { return Point2d { x = self[0], y = self[1] }}
+fn pt(self: [num; 3]): Point3d { return Point3d { x = self[0], y = self[1], z = self[2] }}
+
+// Used as
+p = pt(0, 0)    // p: Point2d
+q = pt(0, 0, 0) // q: Point3d
+```
+
+From the perspective of naming and overriding, there is no difference between a function declared in a module and one imported into a module. In the future, we might allow functions to be imported or declared in smaller scopes, for example in the body of a function or block (which are inside a wider scope such as a module or function body or block). Such names should not be visible outside the narrower scope (we might want some export mechanism, but in that case the names would be referred to similarly to field access on objects, rather than as bare names). The remaining question is whether a name in an inner scope can hide a name in an outer scope. I believe they probably should and that this should be integrated with the rules around `self` types, however, I don't think this is important to finalise now.
+
+[^contra]: extension: non-`self` parameters could be allowed to be contravariant, including removing optional arguments or making them non-optional (with or without optional type).
+
+### Dispatch
+
+A function name whether imported from a module or defined locally, defines *a set of functions*. Likewise, referring to a function via a module (`moduleName.fnName`) references a set of functions. Passing or storing a function may refer to a specific function (`a = fn(...) { ... }`) or a set (`fn foo(...) { ... }; a = foo`); in the former case, no dispatch rules are applied (the latter case needs some thought and I would not support it initially).
```

Comment:
I don't understand how `fn foo(...) { ... }; a = foo` is a set of functions, the way I see it there's only 1 function here, it just has two names. I'm confused!